Adds dynamic component support (aka "hierarchical components") with entry points to ACS, adds SB3 PPO agent and ports Acme R2D2 agent

andyk opened the pull request:
Fixes #127, Fixes #146, Fixes #147

- **This PR breaks the registry functionality!**
- Makes it easier to compose agents via local components
- Adds acme_r2d2 and sb3_agent to example_agents directory
- Acme R2D2 agent ported to new dynamic component graph system
- Acme R2D2 successfully runs w/ the following commands:
    - evaluate entry-point: `agentos run agent --entry-point evaluate --param-file parameters.yaml`
    - learn: `agentos run agent --entry-point learn --param-file parameters.yaml`
    - OR use some command line args for evaluate: `agentos run agent --entry-point learn --param-file parameters.yaml -Pnum_episodes=100`
- Simple port of SB3 PPO, which you can run with:
    - evaluate agent: `agentos run agent`
    - learn: `agentos run agent --entry-point learn`
- Replaced the use of shared_data by Acme R2D2 components with direct access to other components
- Evolve param system for new component system. Support params at command line w/ `-P` flag and also a param file w/ `--param-file` flag (command line flags override args specified in the file)
- Moves the metrics tracking and checkpointing functionality from runtime.py into `agento.Agent` since not all components will handle that functionality the same way, so it makes more sense for it to be associated with the optional/example agent component we are providing in agentos.Agent. 
	- Now the runtime is responsible for managing and loading parameters, allowing the user to call entry-points, initing agent projects, installing components, and interacting with the registry, while all tracking and artifact logging is handled inside of components (probably by using libraries)
- Renames the default entry point from "run" to "evaluate" (to semantically separate the concept of agentos "run"ing an entry point vs the purpose of the entry point being to evaluate the performance of an agent.
- Makes `reset()` an entry_point in `core.Agent`, though this isn't tested yet
- TODO: add a way for AgentOS cli to show the params supported by an entry-point by inspecting the component function signatures.
- NOTE: parameters from the CLI via `-P` flag are all sent to component entry point function) as strings, so the component entry point has to cast them to other values if necessary.
- TO DISCUSS: Should we rename the agentos "run" command to be "exec", and `run_component` to be `exec_component` (suggested by Beth)?

nickjalbert reviewed (changes requested):
Thanks for submitting this!  Lots of changes and I'm excited to try it out 😄.  I was able to get both your R2D2 and SB3 examples running on my box!  I propose that we don't worry too much about the broken stuff (e.g. registry, `agentos init`), merge this, and then dig out.

I've left a few minor inline comments, but here are the things I think we should change before merging:

* Github is showing you added a submodule at  `example_agents/acme_r2d2/.aos/acme_r2d2_policy`.  I'm guessing this may have been unintentional.  If so, let's kill it.
* I propose that we remove the `.aos/` folders from the example agents because they contain a bunch of binary files and other clutter.  We should make the runtime recreate the expected directory structure if it does not exist (I believe we already have a function for that).
* I needed to `pip install attrdict` before I could run the R2D2 agent.  Maybe that should be added to requirements.  Same story with the SB3 agent and `pip install stable_baselines3`.
* We can probably remove my personal `.bash_rc` and `.bash_profile` that migrated over from `acme_test` repo and the Dockerfile entries that copy them into the Docker image.
* Build currently isn't green (looks like you need to run `python scripts/format_code.py`).  Maybe we should just skip any tests we're trying to run until we dig out?

Those are the main things I'm thinking for pre-merge.  Let me know if you feel differently on any of them!

Two higher-level impressions that we should mull (but maybe not worry about solving before merging this PR, IMO):

* One trade-off with making the Agent structure more flexible is that now it's difficult to land in a new agent directory and figure out how to use it because the developer can choose to implement whatever entry points they want.   This, I think, reduces the value of knowing the AgentOS framework because there's no standard approach to doing things that will save you time when you're looking at an unfamiliar project.  Maybe this won't actually be a problem if there's a README or similar.  We also might consider listing entry points in the component spec.
* The SB3 port is super minimal!  The R2D2 port also looked similar in the initial pass (and later passes didn't add functionality, just refactored it).  I think we should consider what the value prop of AgentOS is beyond a simple Python or shell script when SB3 and Acme (and presumably RLlib etc) have a fairly well-encapsulated interface.  Tracking could be an added benefit, but it seems like either way (AgentOS or a shell script) we'll have to go digging into the guts of SB3 to instrument things appropriately.  Ultimately, I just want to make sure that we're overcomplicating this.

Things we can discuss next time we meet!  Anyway, nice work (I'm quite happy you were able to eliminate `shared_data` from the R2D2 port) and let me know if you want me to help with any pre-merge stuff!

andyk commented on example_agents/sb3_agent/agent.py:
```diff
@@ -0,0 +1,45 @@
+    to AgentOS users (e.g., via the CLI or runtime client API).
+    """
+
+    entry_points = ["run", "learn"]
```
Nope. Removed it.

andyk commented:
>Github is showing you added a submodule at example_agents/acme_r2d2/.aos/acme_r2d2_policy. I'm guessing this may have been unintentional. If so, let's kill it.

Yep, was unintentional. Removed it.

>I propose that we remove the .aos/ folders from the example agents because they contain a bunch of binary files and other clutter. We should make the runtime recreate the expected directory structure if it does not exist (I believe we already have a function for that).

Done. I also moved the tracking functionality out of runtime into an `agentos.tracking.Tracker` class.

>I needed to pip install attrdict before I could run the R2D2 agent. Maybe that should be added to requirements. Same story with the SB3 agent and pip install stable_baselines3.

Done.

>We can probably remove my personal .bash_rc and .bash_profile that migrated over from acme_test repo and the Dockerfile entries that copy them into the Docker image.

Done.

Build currently isn't green (looks like you need to run python scripts/format_code.py). Maybe we should just skip any tests we're trying to run until we dig out?

Done.

andyk commented on agentos/runtime.py:
```diff
@@ -1,61 +1,67 @@
-        )
-        all_steps.append(steps)
+    :param component_spec_file: file containing this component's spec
+        (i.e. 'component spec').
```
I cleaned up the comment by removing "(i.e. 'component spec')."

andyk commented on agentos/runtime.py:
```diff
@@ -1,61 +1,67 @@
+                entry_point_params = fully_qualified_params[component_name][
+                    entry_point
+                ]
+            except KeyError:
```
It means that no params were provided for this entry point of this component.

andyk commented on example_agents/acme_r2d2/requirements.txt:
```diff
@@ -0,0 +1,126 @@
+distro==1.4.0
+dm-acme==0.2.1
+dm-env==1.5
+dm-reverb==0.3.1
```
I think adding the extras specifier in the requirements.txt like this should address it, no?

andyk replied:
(not sure why it isn't labeling this as "Outdated")

andyk commented:
Ok @nickjalbert I think I've addressed all of your current specific points of feedback!

andyk commented on example_agents/acme_r2d2/config.py:
```diff
@@ -0,0 +1,35 @@
+#  Writes out an example config file for this component's local and global
```
I removed this file.

andyk commented on example_agents/acme_r2d2/Dockerfile:
```diff
@@ -0,0 +1,46 @@
+
+# House keeping
+COPY ./agentos/example_agents/acme_r2d2/docker/usr-config/.bashrc $CONTAINER_HOME/.bashrc
+COPY ./agentos/example_agents/acme_r2d2/docker/usr-config/.bash_profile $CONTAINER_HOME/.bash_profile
```
Removed them.

andyk replied:
Note that this new docker stuff is not really tested too much; though I am Docker on my MacBook (non-M1) for dev.

nickjalbert commented:
LGTM!
